adding env variable EnableLabelPrediction

MaximilianoUribe opened the pull request:
#### What type of PR is this?
/kind bug /kind flake

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
Added a new environment variable for label prediction, allowing self-hosted users to operate without relying entirely on input labels.

#### Which issue(s) this PR fixes:
Conflicts labeling upstream.
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Adding this label modifies the behavior of scaling from zero nodes. By default, this functionality is enabled, but it can be overridden if needed—primarily to avoid conflicts with existing AKS labels. To use label prediction instead of relying solely on input labels, self-hosted users can configure the EnableLabelPredictionsOnTemplate setting, which remains the recommended approach.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

comtalyst commented on cluster-autoscaler/cloudprovider/azure/azure_template.go:
```diff
@@ -324,45 +324,47 @@ func processVMSSTemplate(template NodeTemplate, nodeName string, node apiv1.Node
+			}
+			// Add ephemeral-storage value
+			if template.VMSSNodeTemplate.OSDisk.DiskSizeGB != nil {
+				node.Status.Capacity[apiv1.ResourceEphemeralStorage] = *resource.NewQuantity(int64(int(*template.VMSSNodeTemplate.OSDisk.DiskSizeGB)*1024*1024*1024), resource.DecimalSI)
```
Maybe we have to keep this one (or any non-labels in this block)?

comtalyst commented on cluster-autoscaler/cloudprovider/azure/azure_template.go:
```diff
@@ -324,45 +324,47 @@ func processVMSSTemplate(template NodeTemplate, nodeName string, node apiv1.Node
-		} else {
-			labels[legacyStorageProfileNodeLabelKey] = "managed"
-			labels[storageProfileNodeLabelKey] = "managed"
+	if enableLabelPrediction {
```
Maybe add a comment here explaining that this attempt is a best-effort to match AKS system labels, and it requires continuous maintenance to keep the parity?

comtalyst commented on cluster-autoscaler/cloudprovider/azure/azure_config.go:
```diff
@@ -103,6 +103,9 @@ type Config struct {
 	EnableFastDeleteOnFailedProvisioning bool `json:"enableFastDeleteOnFailedProvisioning,omitempty" yaml:"enableFastDeleteOnFailedProvisioning,omitempty"`
+
+	// EnableLabelPrediction defines whether to enable self-hosted label prediction for nodes
+	EnableLabelPrediction bool `json:"enableLabelPrediction,omitempty" yaml:"enableLabelPrediction,omitempty"`
```
nit: The name `EnableLabelPredictionsOnTemplate` might be a better in terms of scoping, but not sure.

comtalyst commented on cluster-autoscaler/cloudprovider/azure/azure_config.go:
```diff
@@ -103,6 +103,9 @@ type Config struct {
 	// EnableFastDeleteOnFailedProvisioning defines whether to delete the experimental faster VMSS instance deletion on failed provisioning
 	EnableFastDeleteOnFailedProvisioning bool `json:"enableFastDeleteOnFailedProvisioning,omitempty" yaml:"enableFastDeleteOnFailedProvisioning,omitempty"`
+
+	// EnableLabelPrediction defines whether to enable self-hosted label prediction for nodes
```
nit: No need to specified self-hosted here? Could be just "defines whether to enable system label predictions for node templates (i.e., scaling from zero)"

comtalyst reviewed (commented):
Overall looking good. Also consider:
- Adding/extending unit tests?
- Try running existing E2Es if not already
- Add a release note, given that this is a change in behavior (e.g., it is default to `false`, thus will remove the ability unless changed)

MaximilianoUribe commented:
> Overall looking good. Also consider:
> 
> * Adding/extending unit tests? (try Copilot)
> * Try running existing E2Es if not already
> * Add a release note, given that this is a change in behavior (e.g., it is default to `false`, thus will remove the ability unless changed)

Addressed comments.

comtalyst commented on cluster-autoscaler/cloudprovider/azure/azure_template.go:
```diff
@@ -324,45 +324,50 @@ func processVMSSTemplate(template NodeTemplate, nodeName string, node apiv1.Node
-		labels[GPULabel] = "nvidia"
-		labels[legacyGPULabel] = "nvidia"
+	// Add ephemeral-storage value
+	if template.VMSSNodeTemplate.OSDisk.DiskSizeGB != nil {
```
Need a nil check `template.VMSSNodeTemplate.OSDisk != nil` too? It was present in the larger if statement that this came from

comtalyst commented on cluster-autoscaler/cloudprovider/azure/azure_template.go:
```diff
@@ -324,45 +324,50 @@ func processVMSSTemplate(template NodeTemplate, nodeName string, node apiv1.Node
-		labels[legacyGPULabel] = "nvidia"
+	// Add ephemeral-storage value
+	if template.VMSSNodeTemplate.OSDisk != nil && template.VMSSNodeTemplate.OSDisk.DiskSizeGB != nil {
+		node.Status.Capacity[apiv1.ResourceEphemeralStorage] = *resource.NewQuantity(int64(int(*template.VMSSNodeTemplate.OSDisk.DiskSizeGB)*1024*1024*1024), resource.DecimalSI)
```
Just a small one: could you add a unit test to validate this as well?

comtalyst commented:
/lgtm
/hold

comtalyst commented:
/lgtm
/approve

MaximilianoUribe commented:
/cherry-pick cluster-autoscaler-release-1.30

comtalyst commented:
/cherry-pick cluster-autoscaler-release-1.33
